torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp

```diff
@@ -160,6 +160,124 @@ class DomainMap {
   ComputeAtMap ca_index_map_;
   std::vector<TensorView*> view_tvs_;
 };
+
+// Merge innermost domains for finding the widest vectorizable
+// size. Return the merged domain or nullptr if no merge is done.
+IterDomain* mergeInnermostDomains(
+    const std::vector<IterDomain*>& domain,
+    int num_merged_domains) {
+  const auto ndims = domain.size();
+  IterDomain* merged_id = nullptr;
+  bool is_merge_done = false;
+  for (const auto i : c10::irange(num_merged_domains)) {
+    auto id = domain.at(ndims - 1 - i);
+    // broadcast and reduction are ignored in schedulePointwise
+    if (id->isBroadcast() || id->isReduction()) {
+      continue;
+    }
+    if (merged_id == nullptr) {
+      merged_id = id;
+    } else {
+      auto id_inner = merged_id;
+      auto id_outer = id;
+      merged_id = IterDomain::merge(id_outer, id_inner);
+      is_merge_done = true;
+    }
+  }
+  return is_merge_done ? merged_id : nullptr;
+}
+
+//! Attempt to expand vectorized domains to contig merged domains.
+size_t expandVectorizationToContigMergedDomains(
+    Fusion* fusion,
+    SchedulerRuntimeInfo& runtime_info,
+    const std::vector<TensorView*> vectorizable_inputs_outputs,
+    TensorView* reference_tv,
+    int break_point,
+    size_t default_word_size) {
+  // Don't vectorize when RNG is used
+  if (fusion->isStochastic() && disableRNGUnrolling()) {
+    return default_word_size;
+  }
+
+  auto ca_map = ComputeAtMap(fusion, ComputeAtMap::MappingMode::INDEX);
```

Comment:
I thought this is not on a hot path. This is only called from `getPointwiseHeuristics`. Would it be a performance concern?

Reply:
`getPointwiseHeuristics` is not on the static shape hot path.

For dynamic shaped cases it's on the hot path since we do need to re-evaluate the heuristic values for the new shapes and so this could slow things down when, say we have input sequences with varying lengths. Caching this could go in a follow up if needed.

Reply:
Oh, I see. Yes, for dynamic shapes, this would be exercised every time, but it would also mean there would be lowering and nvrtc compilation times, so compared to those times, building `ComputeAtMap` here would be unlikely a major overhead.